### PR TITLE
Workaround "ImageQueryLod"

### PR DESCRIPTION
**Pull request by AMD-dwang**
When UV width equals 0, the result will return 0, but we expect the value is
-22.0 to -infinity. we perform derivative operations on the XY of UV coordinate
 respectively, if both of result are zero, we will return -FLT_MAX for LOD.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Review by kuhar — COMMENTED** [review]
Can we have a LIT test for this?

**Comment by AMD-dwang** [comment]
> 

This workaround is only related to the input data, and has nothing to do with the shader itself.
 only when we use same texture coordinates (fwidth(UV) == 0) we can find the difference. 
and cts tests will be added. Thanks.

**Comment by kuhar** [comment]
@AMD-dwang 
> This workaround is only related to the input data, and has nothing to do with the shader itself.
only when we use same texture coordinates (fwidth(UV) == 0) we can find the difference.
and cts tests will be added. Thanks.

Sorry, I don't follow. This pass takes SPIR-V as input and outputs LLVM IR. Why can't we have a LIT test that makes sure that the code that comes out of the translator has the expected LLVM instructions?

With CTS the problem is that you need a GPU to execute the compiled code, which we don't have on generic CI bots and most development workstations. There's also no clear connection between the code and CTS tests. This makes it difficult to tell if the code is doing what is was designed to do or jus happens to work, or if our in-development changes cause breakage.

**Comment by AMD-dwang** [comment]
> @AMD-dwang
> 
> > This workaround is only related to the input data, and has nothing to do with the shader itself.
> > only when we use same texture coordinates (fwidth(UV) == 0) we can find the difference.
> > and cts tests will be added. Thanks.
> 
> Sorry, I don't follow. This pass takes SPIR-V as input and outputs LLVM IR. Why can't we have a LIT test that makes sure that the code that comes out of the translator has the expected LLVM instructions?
> 
> With CTS the problem is that you need a GPU to execute the compiled code, which we don't have on generic CI bots and most development workstations. There's also no clear connection between the code and CTS tests. This makes it difficult to tell if the code is doing what is was designed to do or jus happens to work, or if our in-development changes cause breakage.

Okay, I will add a LIT test, thanks.

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 1]
```diff
@@ -6336,6 +6336,32 @@ Value *SPIRVToLLVM::transSPIRVImageQueryLodFromInst(SPIRVInstruction *bi, BasicB
     result = ConvertingSamplerSelectLadderHelper(result, convertingSamplerIdx, createImageGetLod);
   }
 
+  // NOTE: This is a workaround. When UV width equals 0, the result return 0, but we expect the value is
+  // -22.0 to -infinity.
+  // Derivative operations are performed on the XY of UV coordinate respectively, if both of result
+  // are zero, we will return -FLT_MAX for LOD.
+  Value *dpdx = getBuilder()->CreateDerivative(coord, false, false);
+  Value *dpdy = getBuilder()->CreateDerivative(coord, true, false);
+  Value *absDpdx = getBuilder()->CreateUnaryIntrinsic(Intrinsic::fabs, dpdx);
```
Comment (amdrexu) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:6345]: Let's try to reduce some local variables
Value *absDpdx = getBuilder()->CreateUnaryIntrinsic(Intrinsic::fabs, getBuilder()->CreateDerivative(coord, false, false));
Value *absDpdy = ...
Remove dpdx and dpdy

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 2]
```diff
@@ -6336,6 +6336,32 @@ Value *SPIRVToLLVM::transSPIRVImageQueryLodFromInst(SPIRVInstruction *bi, BasicB
...
+  Value *absDpdy = getBuilder()->CreateUnaryIntrinsic(Intrinsic::fabs, dpdy);
+  Value *add = getBuilder()->CreateFAdd(absDpdx, absDpdy);
```
Comment (amdrexu) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:6347]: add -> absDdpxPlusDpdy

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 3]
```diff
@@ -6336,6 +6336,32 @@ Value *SPIRVToLLVM::transSPIRVImageQueryLodFromInst(SPIRVInstruction *bi, BasicB
...
+  Value *cmp = getBuilder()->CreateFCmpOEQ(add, ConstantFP::get(coord->getType(), 0.0));
```
Comment (amdrexu) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:6348]: cmp -> maybeZero

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 4]
```diff
@@ -6336,6 +6336,32 @@ Value *SPIRVToLLVM::transSPIRVImageQueryLodFromInst(SPIRVInstruction *bi, BasicB
...
+
+  Value *con = getBuilder()->getInt1(true);
```
Comment (amdrexu) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:6350]: con -> isZero
getInt1(true) -> getTrue()

File: llpc/translator/lib/SPIRV/SPIRVReader.cpp [hunk 5]
```diff
@@ -6336,6 +6336,32 @@ Value *SPIRVToLLVM::transSPIRVImageQueryLodFromInst(SPIRVInstruction *bi, BasicB
...
+  if (cmp->getType()->isVectorTy()) {
+    auto vec = cast<FixedVectorType>(cmp->getType());
```
Comment (amdrexu) [inline llpc/translator/lib/SPIRV/SPIRVReader.cpp:6352]: Could merge these two statements
for (unsigned elemIdx = 0; elemIdx < cast<FixedVectorType>(maybeZero->getType())->getNumElements(); elemIdx++) {

Comment (AMD-dwang) [reply llpc/translator/lib/SPIRV/SPIRVReader.cpp]: Fixed.

**Review by amdrexu — APPROVED** [review]

**Pull request merged**